crates/pera-core/src/authority.rs

```diff
@@ -1522,6 +1533,28 @@ impl AuthorityState {
         Ok(())
     }
 
+    async fn handle_mpc_events(
+        &self,
+        inner_temporary_store: &InnerTemporaryStore,
+        effects: &TransactionEffects,
+        epoch_store: &Arc<AuthorityPerEpochStore>,
+    ) -> anyhow::Result<()> {
+        if !self.is_validator(epoch_store) {
+            return Ok(());
+        }
+        let status = match &effects {
+            TransactionEffects::V1(effects) => effects.status(),
+            TransactionEffects::V2(effects) => effects.status(),
+        };
+        if status.is_err() {
+            // If the transaction failed, we don't need to handle MPC events.
+            return Ok(());
+        }
+        self.mpc_service
+            .lock()
+            .handle_mpc_events(&inner_temporary_store.events.data)
```

Comment:
something to consider

```suggestion
        if status.is_ok() {
            self.mpc_service
                .lock()
                .handle_mpc_events(&inner_temporary_store.events.data)
        } else {
            // If the transaction failed, we don't need to handle MPC events.
            Ok(())
        }
```